README.md
A little utility that presents current time at given address.

Uses [Google Maps API][maps], so you'll need network access.

[maps]: https://developers.google.com/maps/documentation/timezone

# Example Usage

    $ ./timeat 'los angeles'
    Los Angeles, CA, USA: Fri May 23, 2014 22:33
    $ ./timeat paris
    Paris, France: Sat May 24, 2014 07:37
    Paris, TX, USA: Sat May 24, 2014 00:37
    Paris, TN 38242, USA: Sat May 24, 2014 00:37
    Paris, IL 61944, USA: Sat May 24, 2014 00:37
    Paris, KY 40361, USA: Sat May 24, 2014 01:37
    $

# Installing

You can either `go get github.com/tebeka/timeat` or download the files from
[github][gh].

[gh]: https://github.com/tebeka/timeat/releases


ChangeLog
2017-12-19 version 2.1.0
* Support API key

2016-12-26 version 2.0.1
* Add main.go to git (doh!)

2016-12-26 version 2.0.0
* timeat/lib
* Moving to github

2015-09-16 version 1.1.1
* Added simple test
* Move ntp code to ntp.go

2014-08-21 version 1.1.0
* Get time from pool.ntp.org

2014-08-21 version 1.0.0
* Bump to 1.0.0
* Small usage fix

2014-07-30 version 0.1.1
* Fix logic error not using UTC


lib/timeat_test.go
package timeat

import (
	"strings"
	"testing"
)

func init() {
	SetAPIKey("")
}

func TestTimeAt(t *testing.T) {
	addr := "sunnyvale"
	tis, err := TimeAt(addr)
	if err != nil {
		t.Fatalf("can't get time at %s - %s", addr, err)
	}
	found := false
	for _, ti := range tis {
		if strings.Contains(ti.Address, "Sunnyvale") {
			found = true
			break
		}
	}

	if !found {
		t.Fatalf("no 'Sunnyvale' in output: %v", tis)
	}
}


lib/ntp.go
package timeat

// NTP client code

import (
	"encoding/binary"
	"net"
	"time"
)

type ntpTime struct {
	Seconds  uint32
	Fraction uint32
}

type ntpMsg struct {
	LiVnMode       byte // Leap Indicator (2) + Version (3) + Mode (3)
	Stratum        byte
	Poll           byte
	Precision      byte
	RootDelay      uint32
	RootDispersion uint32
	ReferenceID    uint32
	ReferenceTime  ntpTime
	OriginTime     ntpTime
	ReceiveTime    ntpTime
	TransmitTime   ntpTime
}

func (t ntpTime) UTC() time.Time {
	nsec := uint64(t.Seconds)*1e9 + (uint64(t.Fraction) * 1e9 >> 32)
	return time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC).Add(time.Duration(nsec))
}

// NTPTime returns the current time (UTC) from NTP
// NTP code adapted from https://github.com/beevik/ntp/blob/master/ntp.go
func NTPTime() (time.Time, error) {
	conn, err := net.Dial("udp", "pool.ntp.org:123")
	if err != nil {
		return time.Time{}, nil
	}
	defer conn.Close()
	conn.SetDeadline(time.Now().Add(2 * time.Second))

	msg := ntpMsg{
		LiVnMode: 0x1B,
	}

	err = binary.Write(conn, binary.BigEndian, &msg)
	if err != nil {
		return time.Time{}, err
	}

	err = binary.Read(conn, binary.BigEndian, &msg)
	if err != nil {
		return time.Time{}, err
	}

	return msg.ReceiveTime.UTC(), nil
}


lib/timeat.go
// package timeat finds current time at address using Google Maps API

package timeat

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"os/user"
	"path"
	"strings"
	"time"
)

const (
	apiBase = "https://maps.googleapis.com/maps/api"
	geoURL  = apiBase + "/geocode/json"
	tzURL   = apiBase + "/timezone/json"

	// Version is the package version
	Version = "2.1.0"
)

var apiKey = ""

type apiLoc struct {
	Address  string `json:"formatted_address"`
	Geomerty struct {
		Location struct {
			Lat float32 `json:"lat"`
			Lng float32 `json:"lng"`
		} `json:"location"`
	} `json:"geometry"`
}

type apiGEOReply struct {
	Status    string   `json:"status"`
	Locations []apiLoc `json:"results"`
}

type apiTZReply struct {
	DST    float32 `json:"dstOffset"`
	Offset float32 `json:"rawOffset"`
	Status string  `json:"status"`
}

// TimeInfo holds local time at an address
type TimeInfo struct {
	Address string    // Address
	Time    time.Time // Local time
}

func findAPIKey() string {
	key := os.Getenv("TIMEIT_API_KEY")
	if key != "" {
		return key
	}

	user, err := user.Current()
	if err != nil {
		return ""
	}

	keyFile := path.Join(user.HomeDir, ".config", "timeat", "api-key.txt")
	data, err := ioutil.ReadFile(keyFile)
	if err != nil {
		return ""
	}

	return strings.TrimSpace(string(data))
}

// SetAPIKey sets the API key
// If key is "" will try to read from TIMEIT_API_KEY environment variable or
// from ~/.config/timeat/api-key.txt
func SetAPIKey(key string) {
	if key == "" {
		key = findAPIKey()
	}

	if key != "" {
		apiKey = key
	}
}

// apiCall calls Google Geo API and populates reply
func apiCall(url string, vals url.Values, reply interface{}) error {
	if apiKey != "" && vals.Get("key") == "" {
		vals.Add("key", apiKey)
	}
	url = fmt.Sprintf("%s?%s", url, vals.Encode())
	resp, err := http.Get(url)
	if err != nil {
		return fmt.Errorf("can't call %s - %s", url, err)
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("bad return code - %s", resp.Status)
	}
	defer resp.Body.Close()
	dec := json.NewDecoder(resp.Body)
	if err := dec.Decode(reply); err != nil {
		return fmt.Errorf("can't decode reply - %s", err)
	}

	return nil
}

// TimeAt return list of location matching address and the local time in each
// of them
func TimeAt(address string) ([]TimeInfo, error) {

	var times []TimeInfo

	vals := url.Values{}
	vals.Add("address", address)
	grep := apiGEOReply{}
	if err := apiCall(geoURL, vals, &grep); err != nil {
		return nil, fmt.Errorf("can't get geo for %s - %s", address, err)
	}

	if grep.Status != "OK" {
		return nil, fmt.Errorf("bad geo status - %s", grep.Status)
	}

	now, err := NTPTime()
	if err != nil {
		now = time.Now().UTC()
	}

	for _, loc := range grep.Locations {
		vals = url.Values{}
		vals.Add("timestamp", fmt.Sprintf("%d", now.Unix()))
		geo := loc.Geomerty.Location
		vals.Add("location", fmt.Sprintf("%f,%f", geo.Lat, geo.Lng))
		tz := apiTZReply{}
		if err := apiCall(tzURL, vals, &tz); err != nil {
			return nil, fmt.Errorf("can't get geo for %s - %s", address, err)
		}
		if tz.Status != "OK" {
			return nil, fmt.Errorf("bad tz status - %s", tz.Status)
		}
		dst, offset := time.Duration(tz.DST), time.Duration(tz.Offset)
		local := now.Add((dst + offset) * time.Second)

		times = append(times, TimeInfo{loc.Address, local})
	}

	return times, nil
}


main.go
/*
Example command line usage

    $ ./timeat 'los angeles'
    Los Angeles, CA, USA: Fri May 23, 2014 22:33
    $ ./timeat paris
    Paris, France: Sat May 24, 2014 07:37
    Paris, TX, USA: Sat May 24, 2014 00:37
    Paris, TN 38242, USA: Sat May 24, 2014 00:37
    Paris, IL 61944, USA: Sat May 24, 2014 00:37
    Paris, KY 40361, USA: Sat May 24, 2014 01:37
    $
*/
package main

import (
	"flag"
	"fmt"
	"os"
	"path"

	"github.com/tebeka/timeat/lib"
)

// die prints error message and aborts the program
func die(format string, args ...interface{}) {
	msg := fmt.Sprintf(format, args...)
	fmt.Fprintf(os.Stderr, "error: %s\n", msg)
	os.Exit(1)
}

func main() {
	var showVersion bool
	var apiKey string
	flag.BoolVar(&showVersion, "version", false, "show version")
	flag.StringVar(&apiKey, "api-key", "", "Google API key (also TIMEIT_API_KEY)")

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "usage: %s ADDRESS\n", path.Base(os.Args[0]))
		flag.PrintDefaults()
	}
	flag.Parse()
	if showVersion {
		fmt.Println(timeat.Version)
		os.Exit(0)
	}

	if flag.NArg() != 1 {
		die("wrong number of arguments")
	}

	timeat.SetAPIKey(apiKey)

	address := flag.Arg(0)
	timesat, err := timeat.TimeAt(address)
	if err != nil {
		die("error: can't get times for %s - %s", address, err)
	}

	if len(timesat) == 0 {
		die("error: no locations found matching %s", address)
	}

	for _, ta := range timesat {
		fmt.Printf("%s: %s\n", ta.Address, ta.Time.Format("Mon Jan 2, 2006 15:04"))
	}
}
